File: src/dynamic_programming/cpp/all_construct/src/all_construct.cpp
#include "all_construct.h"
#include <functional>
#include <unordered_map>

std::vector<std::vector<std::string>>
allConstructBasic(const std::string &target,
                  const std::vector<std::string> &word_bank) {

  std::function<std::vector<std::vector<std::string>>(const std::string &)>
      recurse = [&](const std::string &target)
      -> std::vector<std::vector<std::string>> {
    if (target.empty()) {
      return {{}};
    }

    std::vector<std::vector<std::string>> result;
    for (const auto &word : word_bank) {
      if (target.find(word) == 0) {
        auto suffix_ways = recurse(target.substr(word.size()));
        for (auto &way : suffix_ways) {
          way.push_back(word);
        }
        result.insert(result.end(), suffix_ways.begin(), suffix_ways.end());
      }
    }

    return result;
  };

  return recurse(target);
}

std::vector<std::vector<std::string>>
allConstructMemo(const std::string &target,
                 const std::vector<std::string> &word_bank) {
  std::unordered_map<std::string, std::vector<std::vector<std::string>>> memo;

  std::function<std::vector<std::vector<std::string>>(const std::string &)>
      recurse = [&](const std::string &target)
      -> std::vector<std::vector<std::string>> {
    if (target.empty()) {
      return {{}};
    }

    if (memo.count(target) > 0) {
      return memo[target];
    }

    std::vector<std::vector<std::string>> result;
    for (const auto &word : word_bank) {
      if (target.find(word) == 0) {
        auto suffix_ways = recurse(target.substr(word.size()));
        for (auto &way : suffix_ways) {
          way.push_back(word);
        }
        result.insert(result.end(), suffix_ways.begin(), suffix_ways.end());
      }
    }

    memo[target] = result;
    return result;
  };

  return recurse(target);
}

std::vector<std::vector<std::string>>
allConstructTable(const std::string &target,
                  const std::vector<std::string> &word_bank) {
  std::vector<std::vector<std::vector<std::string>>> table(target.size() + 1);
  table[0] = {{}};

  for (size_t i = 0; i <= target.size(); ++i) {
    for (const auto &word : word_bank) {
      if (target.substr(i, word.size()) == word) {
        auto new_combination = table[i];
        for (auto &combination : new_combination) {
          combination.push_back(word);
        }
        table[i + word.size()].insert(table[i + word.size()].end(),
                                      new_combination.begin(),
                                      new_combination.end());
      }
    }
  }

  return table[target.size()];
}

File: src/dynamic_programming/python/all_construct/src/all_construct.py
from typing import List


def all_construct_basic(target: str, word_bank: List[str]) -> List[List[str]]:
    """
    Find all ways to construct target from words in word_bank.

    Uses basic recursion without memoization.

    Args:
        target: The target string to construct.
        word_bank: List of words that can be used (can be reused).

    Returns:
        List of all possible combinations of words that construct target.
    """

    def recurse(target: str) -> List[List[str]]:
        if not target:
            return [[]]

        result: List[List[str]] = []
        for word in word_bank:
            if target.startswith(word):
                suffix_ways = recurse(target[len(word) :])
                result.extend([way + [word] for way in suffix_ways])

        return result

    return recurse(target)


def all_construct_memo(target: str, word_bank: List[str]) -> List[List[str]]:
    """
    Find all ways to construct target from words in word_bank.

    Uses memoization for efficient computation.

    Args:
        target: The target string to construct.
        word_bank: List of words that can be used (can be reused).

    Returns:
        List of all possible combinations of words that construct target.
    """

    def recurse(target: str, memo: dict) -> List[List[str]]:
        if not target:
            return [[]]

        if target in memo:
            return memo[target]

        result: List[List[str]] = []
        for word in word_bank:
            if target.startswith(word):
                suffix_ways = recurse(target[len(word) :], memo)
                result.extend([way + [word] for way in suffix_ways])

        memo[target] = result
        return result

    return recurse(target, {})


def all_construct_table(target: str, word_bank: List[str]) -> List[List[str]]:
    """
    Find all ways to construct target from words in word_bank.

    Uses tabulation (bottom-up DP) for efficient computation.

    Args:
        target: The target string to construct.
        word_bank: List of words that can be used (can be reused).

    Returns:
        List of all possible combinations of words that construct target.
    """
    table: List[List[List[str]]] = [[] for _ in range(len(target) + 1)]
    table[0] = [[]]

    for i in range(len(target)):
        for word in word_bank:
            if target[i : i + len(word)] == word:
                new_combinations = [combination + [word] for combination in table[i]]
                table[i + len(word)].extend(new_combinations)

    return table[-1]
